### DISCUSSION TITLE
ILI9488 still included in current release 12?

### DISCUSSION THREAD

**Discussion by augea** [arendst/Tasmota#18353] [Q&A]
I'm tying to compile my own version of tasmota32.
I added USE_SPI ,  USE_DISPLAY_ILI9488 and USE_DISPLAY in user_config_override.
I'm able to configure SPI but I'm not able to active displaymode 8.
I check the source code but in subfolder tasmota_xdsp_display is no ILI9488 included.
Is this correct or is ILI9488 obsolte with current Tasmota version 12?

**Comment by sfromis** [comment]
Did you config SPI with the `ILI9488 CS` ? 
In general, those DisplayModel is not something you activate, more like an indication that Tasmota was able to follow your config for the display.

However, I'd suggest instead using the [Universal Display driver](https://tasmota.github.io/docs/Displays/#universal-display-driver), where you config a display.ini file with the characteristics of the display, without Tasmota needing to have code for each display type. And there are [display descriptor](https://github.com/arendst/Tasmota/tree/development/tasmota/displaydesc) files present for ILI9488. With uDisplay, you should not use `ILI9488 CS`.

**Reply by augea** [reply]
@sfromis Thank you for you helpful feedback even I don't you answer fully.
ITL9488 CS means the separate driver next to the universal driver -right? CS is in this case not chip select.

I was able to use the universal driver but still got no access to the display.
A little research brought the answer.
Tasmota only supports 3-wire SPI.
I thought this doesn’t matter since the chip supports both modes but I was wrong.
I checked the datasheet of the ILI9488 and found that the interface must be programmed with IM2, IM1, IM0 pins which are exposed only on the 40-wire connector. On my board these pins are all connect to VCC and thus provides only a 4-wire SPI. I assume it is easy to provide a jumper to change the mode but this is not implemented on all the boards I saw. I checked Aliexpress but all boards provide a 4-wire SPI.
Only 40-Pin connectors (only plastic foil connector or boards for parallel access) can access the pins for programming the system interface.
It turns out that the 3-wire SPI is very exotic for boards with pin-headers. I also checked the recommend seller (buydisplay) were I can get 3-wire SPI boards but it is more complicated since customs and tax is not done by the seller.
My question is why is only the exotic 3-wire SPI supported by Tasmota. Is this also true for universal driver? Can the universal driver (display.ini) be adapted to support 4-wire SPI?

**Reply by sfromis** [reply]
Without uDisplay, `ILI9488 CS` is very much chip select, instead of `SPI CS`.

Anyway, as you post no precise details of you how you set it up, hard to guess if you are not doing something "quite right". Additional common pitfalls is the need to also power on the display (as seen from Tasmota, use the toggle), and making sure that the dimmer level is bright enough to see.

**Reply by augea** [reply]
@sfromis Let me give you more details:
1.	I compiled Tasmota with SPI and universal displaydriver and uplaoded to ESP32WROOM module.
2.	I configured the ports
GPIO13 to SPI MOSI (1)
GPIO14 to SPI CLK (1)
GPIO15 to CS (1)
GPIO18 to Option A (3)
GPIO19 to Relais (1)
GPIO23 to MISO (1)
3.	I copied ILI9488_display.ini to display.ini and uploaded the file to the fs (root) on ESP32
4.	I connected the panel
3.3V to VCC
GND to GND
GPIO15 to CS
GPIO14 to SCK
GPIO13 to SDI
GPIO23 to SDO
GPIO19 to BL
5.	I set displaymodel to 17 (console)
6.	I restarted the ESP32
7.	When I toggle GPIO19 (relay)  backlight is switched on/off
8.	I set displaydimmer 50 
9.	I set displaytext [s1l1c1]Temp: but nothing happens
10.	Displaymode 1 to get date and time but nothing happens on the display
This the log:
`00:00:00.003 HDW: ESP32-D0WDQ6 (PSRAM)
00:00:00.028 UFS: FlashFS mounted with 304 kB free
00:00:00.048 CFG: Loaded from File, Anzahl 130
00:00:00.058 QPC: Count 1
00:00:00.065 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO23(MISO)
00:00:00.171 BRY: Berry initialized, RAM used=4041 bytes
00:00:00.188 TFS: File '.drvset003' not found
00:00:00.188 CFG: Energy use defaults as file system not ready or file not found
00:00:01.209 DSP: Configured display 'ILI9488'
00:00:01.227 Projekt tasmota - Tasmota Version 12.4.0(tasmota)-2_0_6(2023.04.07 08:24:13)
00:00:03.202 WIF: verbinden mit AP1 RAPTOR1 Channel 1 BSSId 34:81:C4:F2:1C:A1 in Modus 11n als tasmota-CCD924-6436...
00:00:09.095 WIF: verbunden
16:53:55.126 HTP: Web-Server aktiv bei tasmota-CCD924-6436 mit IP-Adresse 192.168.1.138
16:53:56.177 QPC: Reset
16:53:59.049 RSL: INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"12.4.0(tasmota)","FallbackTopic":"cmnd/DVES_CCD924_fb/","GroupTopic":"cmnd/tasmotas/"}}
16:53:59.064 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-CCD924-6436","IPAddress":"192.168.1.138","IP6Global":"2003:df:f71b:6800:eb8:15ff:fecc:d924","IP6Local":"fe80::eb8:15ff:fecc:d924"}}
16:53:59.078 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":76}}
16:53:59.091 RSL: RESULT = {"POWER1":"ON"}
16:53:59.092 RSL: POWER1 = ON
16:53:59.093 RSL: RESULT = {"POWER2":"ON"}
16:53:59.094 RSL: POWER2 = ON
16:54:03.509 RSL: STATE = {"Time":"2023-04-07T16:54:03","Uptime":"0T00:00:15","UptimeSec":15,"Heap":167,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":1071,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":43},"POWER1":"ON","POWER2":"ON","Wifi":{"AP":1,"SSId":"RAPTOR1","BSSId":"34:81:C4:F2:1C:A1","Channel":1,"Mode":"11n","RSSI":90,"Signal":-55,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:54:16.909 CMD: displaydimmer 50
16:54:16.916 RSL: RESULT = {"DisplayDimmer":50}
16:54:29.862 CMD: displaytext [s1l1c1]Temp:
16:54:29.866 RSL: RESULT = {"DisplayText":"[s1l1c1]Temp:"}
16:54:37.904 CMD: displaymode 1
16:54:38.815 RSL: RESULT = {"DisplayMode":1}`

**Reply by augea** [reply]
I also tried different ports, but I had no luck.
Can somebody confirm, that the uDisplay is capable of handling 4-wire SPI on ILI4988 displays? 
I would think that universal should be capable of since the descriptor file can handle any display independent of the interface.
So only interface (I2C, SPI, parallel) and registers should be relevant in my opinion. 
Since I used the unedited version of ILI4988 the SPI parameters should come from port definition.
Maybe uDisplay is only working with LVGL and Berry?
I’m thankful for any hint.

**Comment by Jason2866** [comment]
ili4988 is supported. But there is not just one variant of ili4988 displays. So your display.ini needs to have the settings your display needs.
If the config is correct, you will see a short splash text `ili4988` on the display. If you don't see "something" is not correct.
For helping detailed info are needed. Link to display where ALL relevant Infos are. AND does the display in general work? Which test sketch works?

**Reply by sfromis** [reply]
Using the similar ILI9341 with uDisplay and `DisplayText` commands instead of LVGL+Berry works without issues.

Make sure to power the display on using the toggle, and have DisplayDimmer high enough to ensure visibility.

**Reply by augea** [reply]
I did not change the preconfigured file for ILI9488. 
I just renamed it and copied it to file system.
I do I have to define the SPI ports manually since * is not working?
I do not see a splash screen.
I also could not test if the display is defective. 
Maybe I will try with a Raspberry if necessary.
It’s the 3.5 display with (resistive) touch.
This is the link to the display on Aliexpress. 
[https://de.aliexpress.com/item/1005004152942022.html](url)
I can confirm that it is 4-wire SPI since I have seen the IM2, IM1, IM0 are all connected to VCC.
The 14 pin connector is pretty much the same as all ili9488 displays.
1.VDD
2. GND
3. CS
4. RST
5. D/C
6. SDI
7. SCK
8. BL
9. SDO
10. TCK
11. TCS
12. TDI
13. TDO
14. PEN

**Comment by Jason2866** [comment]
Could be i have this display in my drawers... The display shown in the link looks familiar.
Will have a look tomorrow, and if it is the same will try to get it going.

**Comment by Jason2866** [comment]
@augea You mentioned no connection for D/C ?

**Reply by augea** [reply]
@Jason2866 Yes, in first test set I thought it is not necessary. 
I did also a second test set and I configured all pins that could be relevant. **But same result.**
In this case I also used more common used SPI ports of ESP32 WROVER. 

GPIO5 SPI CS <-> CS 
GPIO18 SPI CLK <-> SLK
GPIO19 SPI MISO <-> SDO 
GPIO23 SPI MOSI <-> SDI 
GPIO22 SPI DC <-> D/C
GPIO33 A3

I did not use RST. Do you think this is necessary?

**Comment by clharting** [comment]
Here is my config, it works -- although I don't see the splash screen...
Display.ini: 

:H,ILI9488,480,320,16,SPI,1,*,*,*,*,*,*,*,10 
:S,2,1,1,0,40,20 
:I 
E0,0F,00,03,09,08,16,0A,3F,78,4C,09,0A,08,16,1A,0F 
E1,0F,00,16,19,03,0F,05,32,45,46,04,0E,0D,35,37,0F 
C0,2,17,15 
C1,1,41 
C5,3,00,12,80 
36,1,48 
3A,1,66 
B0,1,80 
B1,1,A0 
B4,1,02 
B6,2,02,02 
E9,1,00 
F7,4,A9,51,2C,82 
11,80 
29,0 
:o,28 
:O,29 
:A,2A,2B,2C,16 
:R,36 
:0,28,00,00,02 
:1,48,00,00,01 
:2,E8,00,00,00 
:3,88,00,00,03 
:P,18 
:i,20,21 
:TS,16 
:B,20,0 
:M,340,3900,300,3800 
# 

Configuration: 

{"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"800":"SPI DC1"},"GPIO3":{"0":"None"},"GPIO4":{"1024":"Display Rst"},"GPIO5":{"6210":"Option A3"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"768":"SPI CS1"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"},"GPIO18":{"736":"SPI CLK1"},"GPIO19":{"672":"SPI MISO1"},"GPIO21":{"0":"None"},"GPIO22":{"0":"None"},"GPIO23":{"704":"SPI MOSI1"},"GPIO25":{"0":"None"},"GPIO26":{"0":"None"},"GPIO27":{"0":"None"},"GPIO32":{"992":"Backlight"},"GPIO33":{"0":"None"},"GPIO34":{"0":"None"},"GPIO35":{"0":"None"},"GPIO36":{"7584":"HallEffect1"},"GPIO39":{"7585":"HallEffect2"}} 

I'm not sure exactly what got it to work, but it seems to need SPI MISO defined and wired. And I wired Reset to GPIO4 (Display Rst).

Custom compile info:
Program Version | 12.4.0(TasmoCompiler-esp32generic)
Build Date & Time | 2023-04-13T17:33:57
Core/SDK Version | 2_0_6/4.4.3.20230129
